sdk/core/src/context.rs

```diff
@@ -1,11 +1,177 @@
+use std::any::{Any, TypeId};
+use std::collections::HashMap;
+use std::sync::Arc;
+
 /// Pipeline execution context.
-#[derive(Clone, Debug, Default)]
+#[derive(Clone, Debug)]
 pub struct Context {
-    _priv: (),
+    pub type_map: HashMap<TypeId, Arc<dyn Any + Send + Sync>>,
+}
+
+impl Default for Context {
+    fn default() -> Self {
+        Self::new()
+    }
 }
 
 impl Context {
+    /// Creates a new, empty Context.
     pub fn new() -> Self {
-        Self::default()
+        Self {
+            type_map: HashMap::new(),
+        }
+    }
+
+    /// Inserts or replaces an entity in the type map. If an entity with the same type was displaced
+    /// by the insert, it will be returned to the caller.
+    pub fn insert_or_replace<E>(&mut self, entity: E) -> Option<Arc<E>>
+    where
+        E: Send + Sync + 'static,
+    {
+        self.type_map
+            .insert(TypeId::of::<E>(), Arc::new(entity))
+            .map(|displaced| displaced.downcast().unwrap())
```

Comment:
nit: this unwrap is fine, but perhaps we want to document why it can never fail as a comment above